tactics/autorewrite.ml

```diff
@@ -439,13 +464,28 @@ let auto_multi_rewrite_with ?(conds=Naive) tac_main lbas cl =
 
 (* Functions necessary to the library object declaration *)
 let cache_hintrewrite (rbase,lrl) =
-  let base = try raw_find_base rbase with Not_found -> HintDN.empty in
-  let max = try fst (Util.List.last (HintDN.find_all base)) with Failure _ -> 0 in
-  let fold i accu r = HintDN.add r.rew_pat (i + max + 1, r) accu in
-  let base = List.fold_left_i fold 0 base lrl in
+  let base = try raw_find_base rbase with Not_found -> empty_rewrite_db in
+  let fold accu r = {
+    rdb_hintdn = HintDN.add r.rew_pat r accu.rdb_hintdn;
+    rdb_order = KNmap.add r.rew_id accu.rdb_maxuid accu.rdb_order;
+    rdb_maxuid = accu.rdb_maxuid + 1;
+  } in
+  let base = List.fold_left fold base lrl in
   rewtab := String.Map.add rbase base !rewtab
 
 let subst_hintrewrite (subst,(rbase,list as node)) =
+  let subst_hint subst hint =
```

Comment:
why move this here?

Reply:
It's an internal function that is only used there, I think it's better to keep the substitution code close to one another otherwise we forget stuff.